Public/Src/Tools/Xldb.Proto/README.md

```diff
@@ -0,0 +1,120 @@
+
+### Restrictions:
+1. The values in the Key must be in order from most general to most specific. This way the prefix search can grab the general information first and then you can scope in as needed with more specific values.
+
+2. There cannot be nested messages within the Key (primitives only). The reason is ProtoBuf will _prepend_ bytes with information related to nested message lengths when serializing, and if these lengths differ between different keys, the prefix search will not work as intended.
+
+An important thing to keep in mind when changing any proto file is that the order of messages matters heavily to keep things forward and backwards compatible. Be careful when adding new fields and deprecating old ones! Furthermore default values (int = 0, bool = false, string = "", message=null, etc) are not serialized on the wire to save space on the wire, and the string representation of these Protobuf objects will not deserialize the default values either (though the instantiated class WILL have the values). 
+
+* Do not be alarmed if the value is missing, it means its either unassigned or it is the default value.
+* To force these values to be printed out in C#, you can `JToken.Parse(JsonConvert.SerializeObject(proto_obj, Formatting.Indented))`
+
+<br>
+
+# ProtoBuf Files:
+
+## Events.Proto
+
+The Events.proto file contains all of the protobuf schemas for the events (i.e. Things parsed from the .xlg file).
+
+The key is defined by the message `EventKey` and the values are the different event messages.
+
+Event Key:
+``` ProtoBuf
+// Event query should be of this format. It also serves as the key to the DB
+message EventKey{
+    // ID of the Event - required
+    ExecutionEventId EventTypeID = 1; 
+
+    // Worker ID, 0 (master) by default
+    uint32 WorkerID = 2;
+
+    // Used for PipExecPerf, ProcExecMonitoring, PipCacheMiss, PipExecStepPerf, ProcessFingerpintComp, DirMembHashed, PipExecDirOut
+    uint32 PipId = 3;
+
+    // PipExecutionStepPerformanceKey, PipExecutionStep
+    PipExecutionStep PipExecutionStepPerformanceKey = 4;
+
+    // ProcessFingerprintComputationKey, FingerprintComputationKind
+    FingerprintComputationKind ProcessFingerprintComputationKey = 5;
+
+    // DirectoryMembershipHashedKey, string representing Directory path
+    string DirectoryMembershipHashedKey = 6;
+
+    // PipExecutionDirectoryOutputKey, string representing Directory path
+    string PipExecutionDirectoryOutputKey = 7;
+
+    // FileArtifactContentDecidedKey, string representing file path
+    string FileArtifactContentDecidedKey = 8;
+
+    // File rewrite count for FileArtifactContentDecidedEvent key
+    int32 FileRewriteCount = 9;
+
+    // Used for DependencyViolatedEvent key
+    uint32 ViolatorPipID = 10; 
+
+    // Used for Events that don't have unique key like StatusReported or for events like PipExecutionStepPerformance that can have multiple keys for the same PipId
+    int32 EventSequenceNumber = 11;    
+}
+```

```

Comment:
what's the value of having it here?

Reply:
talked about fixing it offline

---
In reply to: [324346781](https://github.com/microsoft/BuildXL/pull/840#discussion_r324346781) [](ancestors = 324346781)